src/devices/sound/mas3507d.h

```diff
@@ -22,48 +28,73 @@ class mas3507d_device : public device_t, public device_sound_interface
 	void i2c_scl_w(bool line);
 	void i2c_sda_w(bool line);
 
-	u32 get_frame_count() const { return total_frame_count - buffered_frame_count; }
+	uint32_t get_samples() const { return decoded_samples; }
+	uint32_t get_status() const { return playback_status; }
+
+	void update_stream() { stream->update(); }
+
+	void reset_playback();
+	void start_playback();
+
+	bool is_started;
 
 protected:
 	virtual void device_start() override;
 	virtual void device_reset() override;
 	virtual void sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs) override;
 
 private:
+	void i2c_nak();
+	bool i2c_device_got_address(uint8_t address);
+	void i2c_device_got_byte(uint8_t byte);
+	void i2c_device_got_stop();
+
+	void mem_write(int bank, uint32_t adr, uint32_t val);
+	void run_program(uint32_t adr);
+	void reg_write(uint32_t adr, uint32_t val);
+
+	void fill_buffer();
+	void append_buffer(std::vector<write_stream_view> &outputs, int &pos, int scount);
+
 	devcb_read16 cb_sample;
 
+	enum {
+		CMD_DEV_WRITE = 0x3a,
+		CMD_DEV_READ = 0x3b,
+
+		CMD_DATA_WRITE = 0x68,
+		CMD_DATA_READ = 0x69,
+		CMD_CONTROL_WRITE = 0x6a
+	};
+
 	enum { IDLE, STARTED, NAK, ACK, ACK2 } i2c_bus_state;
 	enum { UNKNOWN, VALIDATED, WRONG } i2c_bus_address;
-
-	std::array<u8, 0xe00> mp3data;
-	std::array<mp3d_sample_t, MINIMP3_MAX_SAMPLES_PER_FRAME> samples;
-	bool i2c_scli, i2c_sclo, i2c_sdai, i2c_sdao;
-	int i2c_bus_curbit;
-	uint8_t i2c_bus_curval;
-	int mp3_count, sample_count, current_rate;
-	u32 total_frame_count, buffered_frame_count;
+	enum { UNDEFINED, CONTROL, DATA_READ, DATA_WRITE, BAD } i2c_subdest;
+	enum { CMD_BAD, CMD_RUN, CMD_READ_CTRL, CMD_WRITE_REG, CMD_WRITE_MEM, CMD_READ_REG, CMD_READ_MEM } i2c_command;
```

Comment:
That's how it was before so I can't answer that.

I made the decision to split off the anonymous enum and the variable itself so in my latest commits. Previously save states for those variables weren't possible due to the anonymous enum, but now it works so I added them to the save states. 

I also stopped being lazy and add the complex types for minimp3 that I skipped when adding save states. I think this should make the mas3507d save state complete now. As part of this change, I decided to update the minimp3 library to latest master since there's about 2-3 years worth of updates. That also gets rid of some old hacks I previously put in the minimp3 library so updating/maintenance should be easier in the future.